### PR TITLE
[GC] dataref typing fixes

### PR DESCRIPTION
**Pull request by kripken**
dataref was not noted as `isRef`, and we also did not handle the LUB for it,
which caused validation errors. After these two fixes, it is possible to add a
testcase that goes through the optimizer.

View without whitespace as the LUB change has a lot of that.

### CODE REVIEW

**Comment by kripken** [comment]
(I enabled auto-merge...)

File: src/wasm/wasm-type.cpp
Comment (tlively) [inline src/wasm/wasm-type.cpp:373]: Maybe it would be worth defining more `static constexpr` values like `_last_basic_type` to make it easier to keep these up to date? For example, we could define `_first_reftype` and `_last_reftype`. On the other hand, this probably wouldn't scale well enough for a large number of categorizations of types.

File: src/wasm/wasm-type.cpp
Comment (tlively) [inline src/wasm/wasm-type.cpp:682]: There should probably be an `isEq` predicate for this case on both `Type` and `HeapType`, just like there is for `isFunction`.

File: src/wasm/wasm-type.cpp
Comment (tlively) [inline src/wasm/wasm-type.cpp:688]: It's probably worth clarifying that `anyref` is enabled by the `gc` feature. (I think for a short time `anyref` had its own feature?)

File: src/wasm/wasm-type.cpp
Comment (aheejin) [inline src/wasm/wasm-type.cpp:373]: Oops, I should've changed it too in #3509, sorry..

**Review by aheejin — APPROVED** [review]
LGTM % @tlively's comments

**Comment by aheejin** [comment]
Oh, it seems I approved the PR saying "LGTM % @tlively's comments", and this was thus automatically merged without having chances to address @tlively's comments... Didn't expect this side effect of auto-merge. Not sure how to address this kind of issues when using the auto-merge.

**Comment by kripken** [comment]
Heh, the same happened to me yesterday... we all need to get used to automerge I guess. It basically means no more "lgtm % comments", which makes me wonder if it's actually saving us any time...

I'll work on a followup for the comments.

**Comment by aheejin** [comment]
Yeah I would rather keep the option of approving with comments... I personally don't want to keep from clicking LGTM because of a typo. But if others prefer the auto merge model I can adjust.

**Comment by tlively** [comment]
I sent GitHub feedback yesterday with the suggestion that you should be able to disable auto-merge from the "finishing your review" pane. Until then, we will have to remember to manually disable auto-merge when we want to LGTM with comments.

**Pull request merged**
